### PR TITLE
Switch client keybinds

### PR DESCRIPTION
**Pull request by awerebea**
This PR introduces the ability to use [switch-client](https://man7.org/linux/man-pages/man1/tmux.1.html#:~:text=SIGTSTP%20(tty%20stop).-,switch%2Dclient,-%5B%2DElnprZ%5D%20%5B) keybinding to define all specified save/restore bindings in a separate "namespace".

This is achieved by adding a new option "named-snapshot-switch-client" that allows the user to specify the client to switch to.

The main changes are in the `named-snapshot.tmux` file where the functions `bind_key`, `bind_prompt_key`, `set_save_bindings`, `set_restore_bindings` and `main` have been updated to handle the switch client context. A new function `set_switch_client_table` has been added to set up the switch client key binding with a named key-table.

The `variables.sh` script has been updated to include the new option "named-snapshot-switch-client" with a default value of an empty string.

### CODE REVIEW

File: named-snapshot.tmux
Comment (spywhere) [inline named-snapshot.tmux:16]: suggestion: could you refactor this into a reusable function call?

```sh
bind_tmux_key() {
  local key_table="$1"
  local key="$2"
  local prompt="$3"
  local action="$4"
  if [ -n "$prompt" ]; then
    tmux bind-key -T "$key_table" "$key" command-prompt -p "$prompt" "run shell '$CURRENT_DIR/scripts/$action-snapshot.sh %1'"
  else
    tmux bind-key -T "$key_table" "$key" run-shell "$CURRENT_DIR/scripts/$action-snapshot.sh $name"
  fi
}

bind_prompt_key() {
  # ...
  local key_table="prefix"
  if [ -n "$switch_client" ]; then
    key_table="$(echo "$switch_client" | cut -d':' -f2)"
  fi
  bind_tmux_key "$key_table" "$key" "$prompt" "$action"
  # ...
}
```

File: named-snapshot.tmux
Comment (spywhere) [inline named-snapshot.tmux:27]: suggestion: could you refactor this into a reusable function call?

With the above newly introduced function, we can also refactor

```sh
bind_key() {
  # ...
  local key_table="prefix"
  if [ -n "$switch_client" ]; then
    key_table="$(echo "$switch_client" | cut -d':' -f2)"
  fi
  bind_tmux_key "$key_table" "$key" "" "$action"
  # ...
}
```

**Review by spywhere — CHANGES_REQUESTED** [review]
Thank you for the MR. There are small suggestions to improve the repetition of the code, the rest looks great to me.

I notice you're simply checking if the switch-client is set or not, then do the different binding call. I'd suggested to use `prefix` as a default key table, then override it if we got a new mapping from switch-client. This way, we don't have to conditionally perform a different call (as it get quite long and might be missed if there are changes to it).

File: named-snapshot.tmux
Comment (spywhere) [reply named-snapshot.tmux:27]: We could even move a check to properly override a key table to a new `bind_tmux_key` altogether to remove even more duplicate code!

File: README.md
Comment (spywhere) [inline README.md:38]: nitpick: would `@named-snapshot-keytable` would be a better naming?

`switch-client` is a command in tmux to switch a client to a different session, with an option to change 'key table'. And what we does in this MR is to support a custom key table.

**Comment by spywhere** [comment]
I'm thinking that it would be even better if we introduced 2 new options. One to set a key bindings to switch to a different key table and another one to specify a name for a key table, gave an option to the user to setup their own switching if needed while also supported built-in.

**Comment by awerebea** [comment]
> I'm thinking that it would be even better if we introduced 2 new options. One to set a key bindings to switch to a different key table and another one to specify a name for a key table, gave an option to the user to setup their own switching if needed while also supported built-in.

Hi! I didn't get it, sorry. We already give the user the opportunity to configure their own switching if necessary.
Moreover, the user needs to explicitly define both the **_switch-client_** key and the **_switch-client_** table name.
And we do this the same way we define all other keymaps in this plugin: we pass a `<key>:<table>` map in the same way as we pass `<key>:<action>` map for save/restore actions.
I really don't think hard-coding any default **_switch-client_** table name is a good idea, especially something like **"prefix"** which seems to be a pretty common name for any _prefix_ and there's a pretty good chance that it's already being used somewhere else in the user's configuration, causing an error.

On the other hand, in the current implementation, the previous default behavior of the plugin has not changed in any way, and there is no chance of causing any conflicts with the existing configuration of current users. There are no breaking changes here. However, if more picky users (like me 😃) want to manage all of their save/restore bindings in a separate table so as not to interfere with or overlap all the other bindings, they can do so by specifying one additional option in their configuration. It doesn't make sense to me to split the current configuration option into multiple options for this single feature.

Regarding your notice about refactoring a piece of code into a separate function call.
The code does not repeat any `tmux bind-key` call with exactly the same set of arguments twice. So the goal of such refactoring can hardly be called a _DRY-improvement_.
Two functions are already responsible for setting up key bindings, and we have only 4 lines of code with such calls, and each of them covers its own unique case.
Of course, we _can_ reduce the code repetition a bit if we build the command line by conditionally expanding the string variable containing the common part of the command in each case, and at the end of both of these functions we just `eval` the resulting `$cmd_line`, but such an improvement, in my opinion, is not worth the time spent, sorry.

After all, this is your project. And the final decision is yours. You can make any changes to this PR as a maintainer if you are interested in it. If not and you do not agree with my arguments, then you can simply reject it. I'm ok with this too and will just continue to use my fork in its current state.

Anyway, thank you for this plugin and your attention.

File: named-snapshot.tmux
Comment (awerebea) [reply named-snapshot.tmux:16]: Replied here: https://github.com/spywhere/tmux-named-snapshot/pull/5#issuecomment-2207154911

File: named-snapshot.tmux
Comment (awerebea) [reply named-snapshot.tmux:27]: Replied here: https://github.com/spywhere/tmux-named-snapshot/pull/5#issuecomment-2207154911

**Comment by spywhere** [comment]
> > I'm thinking that it would be even better if we introduced 2 new options. One to set a key bindings to switch to a different key table and another one to specify a name for a key table, gave an option to the user to setup their own switching if needed while also supported built-in.
> 
> Hi! I didn't get it, sorry. We already give the user the opportunity to configure their own switching if necessary. Moreover, the user needs to explicitly define both the **_switch-client_** key and the **_switch-client_** table name. And we do this the same way we define all other keymaps in this plugin: we pass a `<key>:<table>` map in the same way as we pass `<key>:<action>` map for save/restore actions. I really don't think hard-coding any default **_switch-client_** table name is a good idea, especially something like **"prefix"** which seems to be a pretty common name for any _prefix_ and there's a pretty good chance that it's already being used somewhere else in the user's configuration, causing an error.
> 
> ...redacted...

Fair point, I'll merge and refactor it as needed then. Thank you once again for your contribution, much appreciated.

**Pull request merged**
